### ISSUE TITLE
Is there really any point in overriding equals() and hashCode() for every class?

### ISSUE THREAD

**Issue by davidchoo12** [nus-cs2103-AY1819S1/forum#97]
As advised by Prof @damithc, I understand overriding equals will require overriding hashCode. I also found the equals method documentation from Java:
> Note that it is generally necessary to override the hashCode method whenever this method is overridden, so as to maintain the general contract for the hashCode method, which states that equal objects must have equal hash codes.

But I find it kinda unnecessary to override equals and hashCode for every class. Is there any reason the AB4 team decide to do that?

**Comment by anubh-v** [comment]
One reason I have seen, so far, is to improve testability for a particular class.
We can control how to check for equality, and then use that to write a test to check the functionality.

**Comment by davidchoo12** [comment]
> One reason I have seen, so far, is to improve testability for a particular class.
> We can control how to check for equality, and then use that to write a test to check the functionality.

But if the method is not used in the program code, is it necessary to test them?

**Comment by damithc** [comment]
> But if the method is not used in the program code, is it necessary to test them?

Note that the use of equals may not be visible directly. For example, some data structures such as Lists use the `equals` method of its elements internally to check equality. e.g., https://docs.oracle.com/javase/7/docs/api/java/util/List.html#contains%28java.lang.Object%29

**Comment by davidchoo12** [comment]
> Note that the use of equals may not be visible directly. For example, some data structures such as Lists use the `equals` method of its elements internally to check equality. e.g., https://docs.oracle.com/javase/7/docs/api/java/util/List.html#contains%28java.lang.Object%29

True, but looking at the instances where equals has been overridden, many of the classes aren't used in a list or such data structures requiring equals. In fact, only `Person`, `XmlAdaptedPerson` and `XmlAdaptedTag` are used in a list. `Person` is dependent on the other model classes, so I guess you can argue that the dependencies should override equals. Even so, I think it is unnecessary to override for the other classes.
![image](https://user-images.githubusercontent.com/7817101/47074309-be6cde80-d22c-11e8-8275-b359d924ddc0.png)
